### DISCUSSION TITLE
Extend Generic math to enums

### DISCUSSION THREAD

**Discussion by shadow-cs** [dotnet/csharplang#5093] [Language Ideas]
Regarding the recent availability of Generic math preview, it would make sense we also add something along the lines of `IEnum` which would implement `IConvertible<long>` (which itself was described as a possible future improvement of generic math in the [announcement](https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/#generic-math)) as well as `I{Span}Parseable` (and possibly others) and implement `IEnum` in `Enum` base type. 

This would further enhance the #104.

**Comment by foxesknow** [comment]
Why `IConvertible<long>` ? This wouldn't handle something like `enum Foo : ulong`

**Reply by shadow-cs** [reply]
Good point, I was thinking about it being generic. I'm not sure what the possibilities are to make sure it doesn't break anything so I went with the most compatible version I could find. It would be cool if we could separate `Flags` from non-flags with some constraint as well.

**Comment by jnm2** [comment]
I asked at https://github.com/dotnet/runtime/issues/112819 about having the runtime automatically implement `IBinaryInteger<self>` on all enums. This would provide capabilities such as `ulong.CreateTruncating(anyEnumValue)`, `TEnum.ReadLittleEndian` and `tEnumValue.WriteLittleEndian`, `tEnumValue.GetByteCount()`, and general usability with any generic numerics logic.

**Reply by tannergooding** [reply]
An issue is that not all enums are integers. They can be bool and some runtimes even allow floating-point

**Reply by jnm2** [reply]
Added a line: 

> This would only apply to enums whose underlying type implements `IBinaryInteger<T>`.

**Reply by ds5678** [reply]
After reading the comments in #8789 about the problems with automatic implementation at run time and compile time, I feel like this would be best addressed with:

* Members on enums (#8987)
* Partial enums (#2669)
* Enums implementing interfaces

From there, a source generator could let developers opt-in to these interfaces with an attribute.
